experimental/ir/lower_symbols.go

```diff
@@ -95,59 +96,79 @@ func newOneofSymbol(o Oneof) {
 		fqn:  o.InternedFullName(),
 		data: arena.Untyped(c.arenas.oneofs.Compress(o.raw)),
 	})
-	c.symbols = append(c.symbols, ref[rawSymbol]{ptr: sym})
+	c.exported = append(c.exported, ref[rawSymbol]{ptr: sym})
 }
 
-// buildImportedSymbols builds a symbol table of every symbol in every
-// transitive import. This will contain symbols that are not visible to this
-// file, but visibility can be tested with f.visibleImports.
+// buildImportedSymbols builds a symbol table of every imported symbol.
+//
+// It also enhances the exported symbol table with the exported symbols of each
+// public import.
 func buildImportedSymbols(f File, r *report.Report) {
-	// Only need to merge from the direct imports's tables, since each
-	// import table is fully transitive.
 	imports := f.Imports()
+	locals := f.Context().exported
+
+	var havePublic bool
+	for sym := range seq.Values(imports) {
+		if sym.Public {
+			havePublic = true
+			break
+		}
+	}
+
+	// Form the exported symbol table from the public imports. Not necessary
+	// if there are no public imports.
+	if havePublic {
+		f.Context().exported = symtabMerge(
+			f.Context(),
+			iterx.Chain(
+				iterx.Of(locals),
+				seq.Map(imports, func(i Import) symtab {
+					if !i.Public {
+						// Return an empty symbol table so that the table to
+						// context mapping can still be an array index.
+						return symtab{}
+					}
+					return i.Context().exported
+				}),
+			),
+			func(i int) File {
+				if i == 0 {
+					return f
+				}
+				return f.Context().imports.files[i-1]
+			},
+		)
+	}
+	dedupSymbols(f, &f.Context().exported, nil)
 
-	f.Context().symbols = slicesx.MergeKeySeq(
+	// Form the imported symbol table from the full import list.
+	f.Context().imported = symtabMerge(
+		f.Context(),
 		iterx.Chain(
-			iterx.Of(f.Context().symbols),
+			iterx.Of(locals),
 			seq.Map(imports, func(i Import) symtab {
-				return i.Context().symbols
+				return i.Context().exported
 			}),
```

Comment:
Could use `f.Context().exported` instead of `locals` and then have the map function below return an empty table for public imports (since we effectively already did that merging work above).

Reply:
Not wrong, fixed. Still need to run dedup twice, though.